sample_formatted_code.do_not_merge.c

```diff
@@ -0,0 +1,562 @@
+// Formatted using
+//   clang-format -i sample_formatted_code.do_not_merge.c
+
+// SortIncludes
+#include <stdio.h>
+#include <assert.h>
+#include <string.h>
+#include "sample.llvm.h"
+
+//--- Line Length ---
+// https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Coding_Style#Line-length
+// ColumnLimit: 80
+// ContinuationIndentWidth: 8
+static int some_really_long_variable_definition_that_is_80_chars_long = 1234567;
+static int some_long_variable_definition_that_wraps_and_continues_at_next_line =
+        1234567890123;
+
+//--- Indent ---
+// https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Coding_Style#Indent
+// IndentWidth: 4
+// AlignAfterOpenBracket
+int DecodeEthernet(ThreadVars *tv, DecodeThreadVars *dtv, Packet *p,
+        uint8_t *pkt, uint16_t len, PacketQueue *pq)
+{
+    SCPerfCounterIncr(dtv->counter_eth, tv->sc_perf_pca);
+
+    if (unlikely(len < ETHERNET_HEADER_LEN)) {
+        ENGINE_SET_INVALID_EVENT(p, ETHERNET_PKT_TOO_SMALL);
+        return TM_ECODE_FAILED;
+    }
+}
+
+//--- Braces ---
+// https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Coding_Style#Braces
+int SomeFunction(void)
+{
+    DoSomething();
+}
+
+void brace_style()
+{
+    if (unlikely(len < ETHERNET_HEADER_LEN)) {
+        ENGINE_SET_INVALID_EVENT(p, ETHERNET_PKT_TOO_SMALL);
+        return TM_ECODE_FAILED;
+    }
+
+    if (this) {
+        DoThis();
+    } else {
+        DoThat();
+    }
+}
+
+//--- Flow ---
+// https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Coding_Style#Flow
+void if_style()
+{
+    // AllowShortIfStatementsOnASingleLine
+    if (a)
```

Comment:
can we require using braces even in such minimal cases? I think Rust forces this, and I like it. It helps avoid the issue that we've probably all encountered:
```c
if (user_is_root)
   do_something();
   do_something_else();
```

Reply:
Not with clang-format as it's only formatting existing code. It does not change/fix code.

Yup, been down that road.

One could use clang-tidy to force that. That could probably be another task.

Reply:
Could you open a ticket for clang-tidy in our redmine? I'm still a bit unclear on what it provides. The docs suggest it has somethings scan-build does, but then some things are different? Interested in learning what it could do for us.

(btw we run scan-build in (private) QA)

Reply:
scan-build does **static analysis** of code and as such does flow-control and path analysis. Use it to help finding bugs.

clang-tidy provides **linter-based checks** only that is mostly used to enforce coding standards. It's only based on the AST afaik. It's fully configurable on what to check for.

clang-tidy can also be used to "fix the code for some checks", i.e. it's convenient to use clang-tidy to upgrade" a code base to certain standards, such as the "shall always use braces" in this example.

See also CodeChecker that could combine both: https://clang-analyzer.llvm.org/command-line.html Have never used that one.

Opened https://redmine.openinfosecfoundation.org/issues/3837